server/src/main/java/io/crate/execution/engine/collect/files/FileInput.java

```diff
@@ -36,12 +39,27 @@
      *                     (and may optionally contain a "file hint" - which is the part after the last slash.)
      *                     a concrete implementation may ignore the file hint.
      * @param uriPredicate predicate that a concrete implementation of FileInput must use to pre-filter the returned uris
+     * @param withClauseOptions
      * @return a list of Uris
      * @throws IOException
      */
-    List<URI> listUris(URI fileUri, URI preGlobUri, Predicate<URI> uriPredicate) throws IOException;
+    List<URI> listUris(URI fileUri, URI preGlobUri, Predicate<URI> uriPredicate, Map<String, Object> withClauseOptions) throws IOException;
 
-    InputStream getStream(URI uri) throws IOException;
+    InputStream getStream(URI uri, Map<String, Object> withClauseOptions) throws IOException;
+
+    /**
+     * An optional re-formatter that will reformat URIs passed to
+     * {@link FileInput#listUris(URI, URI, Predicate, Map)}, and {@link FileInput#getStream(URI, Map)}.
+     * It is needed only if the getters of the URIs do not return the expected substrings.
+     * @return a re-formatting function
+     */
+    Function<String, URI> uriFormatter();
```

Comment:
```suggestion
    URI parseURI(String uri);
```

No need to wrap it in a `Function`